### DISCUSSION TITLE
[@types/node] ReadableStream type definition here is not compatible with TypeScript's

### DISCUSSION THREAD

**Discussion by suerta-git** [DefinitelyTyped/DefinitelyTyped#65542] [Issues with a @types package] [answered]
This will fail since types are incompatible

```ts
const response = await fetch('xxx')
stream.Readable.fromWeb(response.body!)  // fail!
```

Error:

```
Argument of type 'ReadableStream<Uint8Array>' is not assignable to parameter of type 'ReadableStream<any>'.
  Type 'ReadableStream<Uint8Array>' is missing the following properties from type 'ReadableStream<any>': values, [Symbol.asyncIterator]
```

Node's version:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/16a2befc8fc262d0e75f4ee486e8276181a9fce1/types/node/stream/web.d.ts#L135

TypeScript's version in `lib.dom.d.ts`: 

https://github.com/microsoft/TypeScript/blob/b53073bf66571b4c99f0a93b8f76fb06b36d8991/lib/lib.dom.d.ts#LL11917C1-L11917C1

[a comment by an automated account is not shown]

**Comment by Semigradsky** [comment]
What's the TypeScript version?

**Reply by peterblazejewicz** [reply]
looks like 4.4 or older, as per history for the file linked:
https://github.com/microsoft/TypeScript/commits/b53073bf66571b4c99f0a93b8f76fb06b36d8991/lib/lib.dom.d.ts
@suerta-git ?

**Reply by suerta-git** [reply]
"typescript": "^5.0.2"
"@types/node": "^20.2.1"

**Comment by Semigradsky** [comment]
Looks like related to https://github.com/microsoft/TypeScript/issues/29867

**Reply by wangzishi** [reply]
Yes, it is indeed related to this problem. the `ReadableStream` interface definition in the lib.dom.ts file generated by typescript is missing `[Symbol.asyncIterator](): AsyncIterableIterator<R>;` which is supported by Nodejs.

**Comment by Semigradsky** [comment]
@julienvincent @wangzishi Please, take a look

**Reply by wangzishi** [reply]
I'll have a look at it this weekend.

**Comment by wangzishi** [comment, marked as answer]
The root cause of this problem is that typescript has removed the type definition that generates [the async
iterator for ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#async_iteration). They also did this because chromium has never implemented this functionality, but
Nodejs already implements this, so the type definition difference is caused.

My personal advice is as follows:

1. if the code only runs in Nodejs, then do the type casting directly: 
```ts
import { default as stream } from 'node:stream'.
import type { ReadableStream } from 'node:stream/web'.

const response = await fetch('xxx')
stream.Readable.fromWeb(response.body as ReadableStream<Uint8Array>)
```

2. If the code needs to take into account the browser's runtime environment, you can add a polyfill:
```js
ReadableStream.prototype[Symbol.asyncIterator] = async function* () {
  const reader = this.getReader()
  try {
    while (true) {
      const {done, value} = await reader.read()
      if (done) return
      yield value
    }
  }
  finally {
    reader.releaseLock()
  }
}
```

<details><summary>中文说明</summary>
<p>
导致这个问题的根本原因是因为 typescript 移除了为 ReadableStream 生成 async
iterator 的类型定义。他们这么做也是因为 chromium 对此功能的一直没有实现。
而 Nodejs 中的已经实现了这个功能，所以导致了两者的类型定义差异。

我的个人有如下建议：

1. 如果的代码只在 Nodejs 中运行，那么直接做类型转换：
```ts
import { default as stream } from 'node:stream';
import type { ReadableStream } from 'node:stream/web';

const response = await fetch('xxx')
stream.Readable.fromWeb(response.body as ReadableStream<Uint8Array>)
```

2. 如果代码需要考虑到浏览器的运行环境，可以添加一个 polyfill：
```js
ReadableStream.prototype[Symbol.asyncIterator] = async function* () {
  const reader = this.getReader()
  try {
    while (true) {
      const {done, value} = await reader.read()
      if (done) return
      yield value
    }
  }
  finally {
    reader.releaseLock()
  }
}
```
</p>
</details>

**Reply by suerta-git** [reply]
Thanks @wangzishi, the reason of the removal of async iterator makes sense. But just making a type casting is not reasonable enough for me.

Considering all my code runs in Node.js environment. Both of `fetch` api and `ReadableStream` class are implemented by Node itself. So the underlying type of `fetch()`'s return value should be definitely same as `ReadableStream` as what `Readable.fromWeb()` expects. And expecting the code snippet can work is also natural for developers.

However typescript prevents it and requires a type casting, it's weird and counterintuitive for me.

Rough ideas:
1. An optional type overriding, makes web apis return node version `ReableStream`s.
2. Is the async iterator necessary for `Readable.fromWeb()`? If no, doesn't require it in parameter any more.

**Reply by TechQuery** [reply]
Recommend a polyfill package: https://github.com/MattiasBuelens/web-streams-polyfill

**Reply by fardolieri** [reply]
I ended up creating a file `stream-polyfill-types.d.ts` and placed it somewhere in my project.

```ts
import type { ReadableStreamAsyncIterator } from 'node:stream/web';

declare global {
  /**
   * This allows us to write code like this:
   *
   * import fs from 'node:fs';
   * import { Readable } from 'node:stream';
   * import { pipeline } from 'node:stream/promises';
   * const response = await fetch('https://example.com/');
   * const webReadableStream = response.body;
   * const writeStream = fs.createWriteStream('myfile');
   * await pipeline(webReadableStream, writeStream); // no type error
   * const nodeReadableStream = Readable.fromWeb(webReadableStream); // no type error
   */
  interface ReadableStream<R = any> extends _ReadableStream<R> {
    values(options?: { preventCancel?: boolean }): ReadableStreamAsyncIterator<R>;
    [Symbol.asyncIterator](): ReadableStreamAsyncIterator<R>;
  }
}
```

**Comment by beorn** [comment]
Would be nice to bump this up in priority for Chrome and WebKit - with streams gaining popularity and being standard, it is a shame you can't use idiomatic `for await` looping over streams:
- https://bugs.chromium.org/p/chromium/issues/detail?id=929585
- https://bugs.webkit.org/show_bug.cgi?id=194379

**Comment by steve-taylor** [comment]
> The root cause of this problem is that typescript has removed the type definition that generates [the async iterator for ReadableStream](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#async_iteration). They also did this because chromium has never implemented this functionality, but Nodejs already implements this, so the type definition difference is caused.
> 
> My personal advice is as follows:
> 
> 1. if the code only runs in Nodejs, then do the type casting directly:
> 
> ```ts
> import { default as stream } from 'node:stream'.
> import type { ReadableStream } from 'node:stream/web'.
> 
> const response = await fetch('xxx')
> stream.Readable.fromWeb(response.body as ReadableStream<Uint8Array>)
> ```
> 
> 2. If the code needs to take into account the browser's runtime environment, you can add a polyfill:
> 
> ```js
> ReadableStream.prototype[Symbol.asyncIterator] = async function* () {
>   const reader = this.getReader()
>   try {
>     while (true) {
>       const {done, value} = await reader.read()
>       if (done) return
>       yield value
>     }
>   }
>   finally {
>     reader.releaseLock()
>   }
> }
> ```
> 
> 中文说明

```
Cannot find module 'node:stream/web' or its corresponding type declarations. ts(2307)
```

**Comment by hexrw** [comment]
This is what worked for me if anyone is wondering (I don't need to polyfill)

```ts
Readable.fromWeb(stream as unknown as ReadableStream<Uint8Array>),
```

**Comment by heroboy** [comment]
in recent typescript, the error is 
```
Argument of type 'ReadableStream<Uint8Array<ArrayBufferLike>>' is not assignable to parameter of type 'ReadableStream<any>'.
  The types of 'getReader(...).closed' are incompatible between these types.
    Type 'Promise<void>' is not assignable to type 'Promise<undefined>'.
      Type 'void' is not assignable to type 'undefined'.ts(2345)
```
web.d.ts
```ts
  interface ReadableStreamGenericReader {
      readonly closed: Promise<undefined>;
      cancel(reason?: any): Promise<void>;
  }
```
lib.dom.d.ts
```ts
interface ReadableStreamGenericReader {
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/ReadableStreamBYOBReader/closed) */
    readonly closed: Promise<void>;
    /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/ReadableStreamBYOBReader/cancel) */
    cancel(reason?: any): Promise<void>;
}

```

**Comment by FunctionDJ** [comment]
I'm confused by this. All I want to do is pass a `stream.Readable.toWeb()` result to a `new Response()` (web globals). Why does this require a type assertion? Is there seriously no simple solution to make the types compatible?
